processing/src/main/java/org/apache/druid/query/RestrictedDataSource.java

```diff
@@ -123,10 +122,9 @@ public boolean isProcessable()
   }
 
   @Override
-  public Function<SegmentReference, SegmentReference> createSegmentMapFunction(Query query)
+  public SegmentMapFunction createSegmentMapFunction(Query query)
   {
-    final Function<SegmentReference, SegmentReference> segmentMapFn = base.createSegmentMapFunction(query);
-    return baseSegment -> new RestrictedSegment(segmentMapFn.apply(baseSegment), policy);
+    return SegmentMapFunction.map(segment -> new RestrictedSegment(segment, policy));
```

Comment:
I understand it's pointless to start from `base.createSegmentMapFunction(query)` because we "know" it will be identity, but IMO it's better form to do it anyway. That way all the datasources work the same way, and there's a degree of future-proofing for restrictions being able to apply to things other than tables.

Reply:
I've done this with `RestrictedDataSource`, but not yet for `RestrictedInputNumberDataSource` since it currently doesn't have a reference to the base datasource like `RestrictedDataSource`, only an inputNumber and the policy. Should we update `RestrictedInputNumberDataSource` to be created with the base of the `RestrictedDataSource`? It looks like this was released in 33, but not sure how well documented stuff is yet.. in other words, if we change it would we need to worry about backwards compatibility?

Reply:
I think for `RestrictedInputNumberDataSource` it's ok that there is no base. The `RestrictedInputNumberDataSource` builds the base into itself, in a way.